tla/ccfraft.tla

```diff
@@ -1065,6 +1076,13 @@ RcvUpdateCommitIndex(i, j) ==
         /\ UpdateCommitIndex(m.dest, m.source, m)
         /\ Discard(m)
 
+RcvProposeVoteRequest(i, j) ==
+    \E m \in MessagesTo(i) :
+        /\ j = m.source
+        /\ m.type = ProposeVoteRequest
```

Comment:
Have you considered adding a conjunct `m.term = currentTerm[i]` to model `r.term == state->current_view` here?

Reply:
No, but I can see that it's a good idea :)

Reply:
Ok, having thought about this a little more, I am now unsure: while this holds if there is no delay, it's clearly something that may not be true if there is a delay in delivery. The term could go up, and the receiving node could receive a stale ProposeVoteRequest that it would then ignore, through no fault of the protocol.

So I think it would have to be `m.term <= currentTerm[i]`.

Reply:
I think an equality check (in the implementation) is preferable to an inequality (currently in this spec). Stale messages will be ignored, which feels safe. Reacting to stale (potentially extremely-delayed) messages may create a spurious election. That's obviously unlive if we had message duplication, and feels undesirable regardless.

Reply:
The implementation is using an [equality check](https://github.com/microsoft/CCF/pull/5697/files#diff-4e61c67c4990874cc31cb6f8cd91e4abe7506492fce0f73a4705d556d6861a42R1732), and ignores the message otherwise.

Perhaps what we want here is:

```
\/ /\ m.term == currentTerm[i]
    /\ Timeout(m.dest)
\/ m.term > currentTerm[i]
```

I have to say that I feel like we're sliding towards creating a duplicate of the implementation of equal verbosity, and I don't know if that's obviously a good thing.

Reply:
I believe all of the `m.term > currentTerm[i]` cases are handled by `RcvUpdateTerm`, so we don't need to include the disjunct here. So to match the implementation, we'd add the single conjunct that @lemmy proposed:

```
...
/\ m.type = ProposeVoteRequest
/\ m.term = currentTerm[i]
/\ Timeout(m.dest)
...
```

Reply:
Ok, I'll simplify it again then.

Reply:
> I have to say that I feel like we're sliding towards creating a duplicate of the implementation of equal verbosity, and I don't know if that's obviously a good thing.

I've made the same observation.  At some point, somebody has to create a more high-level spec reusing on inspired by https://github.com/muenchnerkindl/AbstractRaft/, https://github.com/will62794/raft-by-refinement, or https://github.com/will62794/endive/blob/ind-tree/benchmarks/AbstractStaticRaft.tla, and show that ccfraft refines it.